test/zimcheck-test.cpp

```diff
@@ -662,8 +664,104 @@ TEST(zimcheck, json_bad_checksum)
+      "    {"                                                               "\n"
+      "      'check' : 'metadata',"                                         "\n"
+      "      'level' : 'ERROR',"                                            "\n"
+      "      'code' : 2,"                                                   "\n"
+      "      'message' : 'Title',"                                          "\n"
+      "      'metadata_type' : 'Title'"                                     "\n"
+      "    },"                                                              "\n"
+      "    {"                                                               "\n"
+      "      'check' : 'metadata',"                                         "\n"
+      "      'level' : 'ERROR',"                                            "\n"
+      "      'code' : 2,"                                                   "\n"
+      "      'message' : 'Description',"                                    "\n"
+      "      'metadata_type' : 'Description'"                               "\n"
+      "    },"                                                              "\n"
+      "    {"                                                               "\n"
+      "      'check' : 'main_page',"                                        "\n"
+      "      'level' : 'ERROR',"                                            "\n"
+      "      'code' : 1,"                                                   "\n"
+      "      'message' : 'Main Page Index stored in Archive Header: 4294967295'," "\n"
+      "      'main_page_index' : '4294967295'"                              "\n"
+      "    },"                                                              "\n"
+      "    {"                                                               "\n"
+      "      'check' : 'redundant',"                                        "\n"
+      "      'level' : 'WARNING',"                                          "\n"
+      "      'code' : 8,"                                                   "\n"
+      "      'message' : 'article1.html and redundant_article.html',"       "\n"
+      "      'path2' : 'redundant_article.html',"                           "\n"
+      "      'path1' : 'article1.html'"                                     "\n"
+      "    },"                                                              "\n"
+      "    {"                                                               "\n"
+      "      'check' : 'url_internal',"                                     "\n"
+      "      'level' : 'ERROR',"                                            "\n"
+      "      'code' : 6,"                                                   "\n"
+      "      'message' : 'The following links:\\n- A/non_existent.html\\n(/A/non_existent.html) were not found in article dangling_link.html'," "\n"
+      "      'links' : '- A/non_existent.html\\n',"                         "\n"
```

Comment:
I'm not sure to understand the commit message telling we cannot iterate on the links in mustache.

But we should have a json array of `links` here instead of a escaped multi-line string.

Reply:
The commit message reflects my intention of the same. I added two commit demonstrating the problem. The desired result is achieved with a hack.

Reply:
It seems we can iterate over a list : https://mustache.github.io/mustache.5.html (section "Non-Empty Lists")
We need a list of object instead of a list of string but we can. And it is what you've use in your two last commits.

> I will consider that once the JSON output becomes less trivial.

Maybe it is time to consider this not trivial ?

Reply:
What do you think of https://github.com/nlohmann/json ?

> With the current approach JSON can be output in a streaming fashion (as each error is generated).

Do we really need that ?
We already wait the end of the process to know all the errors and then we report them.
And if we really need it, we can create a json object per error and stream one object after the other.

Reply:
> What do you think of https://github.com/nlohmann/json ?

It looks cute.

> 
> > With the current approach JSON can be output in a streaming fashion (as each error is generated).
> 
> Do we really need that ?
> We already wait the end of the process to know all the errors and then we report them.

If we expect zimcheck to run to an end on a huge ZIM file potentially producing a lot of errors (caused by a buggy scraper or a new check in zimcheck), then we better not accumulate all the errors in memory before printing them. Alternatively we can have a limit on each type of errors.

> And if we really need it, we can create a json object per error and stream one object after the other.

Then we will have to output the enclosing JSON object in a custom way like in the current approach.

---

I don't think that utilizing a json library now will pay off in the short term, but I am fine with doing that and then replacing it with even a better one every time such is found.

Reply:
> Then we will have to output the enclosing JSON object in a custom way like in the current approach.

Yes, but the home implemented json would simply be a opening/close of a array/dict.
All the complex part (complex object/list, value escaping, ...) would be made by the well known and tested library.